File: 30_day_challenge_2020_December/1010_pairs_of_songs_day08.py
################################################################
# Author            : yiorgosynkl (find me in Github: https://github.com/yiorgosynkl)
# Date created      : 20201208
# Problem link      : https://leetcode.com/problems/pairs-of-songs-with-total-durations-divisible-by-60/
################################################################

class Solution:
    # def numPairsDivisibleBy60(self, time: List[int]) -> int:
    #     mods = [0]*60
    #     for t in time:
    #         mods[t % 60] += 1
    #     ans = sum(mods[t]*mods[60-t] for t in range(1,30)) + mods[30]*(mods[30]-1)//2 + mods[0]*(mods[0]-1)//2
    #     return ans 

    # @lee215
    def numPairsDivisibleBy60(self, time):
        c = [0] * 60
        res = 0
        for t in time:
            res += c[-t % 60]
            c[t % 60] += 1
        return res

File: 30_day_challenge_2020_December/1010_pairs_of_songs_day08.cpp
/****************************************************************
* Author            : yiorgosynkl (find me in Github: https://github.com/yiorgosynkl)
* Date created      : 20201208
* Problem link      : https://leetcode.com/problems/pairs-of-songs-with-total-durations-divisible-by-60/
****************************************************************/

class Solution {
public:
    int numPairsDivisibleBy60(vector<int>& time) {
        vector<int> c(60);
        int res = 0;
        for (int t : time){
            res += c[(60 - t % 60) % 60]; // second % for special case where t = 0 => 60 - t = 60
            // res += t > 0 ? c[60 - t % 60] : c[0];
            c[t % 60]++;
        }
        return res;
    }
};